File: python/hql/hql/expr.py
"""HQL v0 filter expressions: == != ^= !^= with and / &&. null literal."""

from __future__ import annotations

from typing import Any, Optional, Union

from hql.row import Row

_OPS = ("!^=", "==", "!=", "^=")


class FilterError(ValueError):
    pass


def parse_filter(expr: str) -> "FilterExpr":
    tokens = _tokenize(expr)
    if not tokens:
        raise FilterError("empty filter")
    return _parse_and(tokens)


def _parse_and(tokens: list[str]) -> "FilterExpr":
    parts: list[Comparison] = [_parse_comparison(tokens)]
    while tokens:
        join = tokens.pop(0)
        if join not in ("and", "&&"):
            raise FilterError(f"expected 'and' or '&&', got {join!r}")
        parts.append(_parse_comparison(tokens))
    if len(parts) == 1:
        return parts[0]
    return And(parts)


def _parse_comparison(tokens: list[str]) -> "Comparison":
    if len(tokens) < 3:
        raise FilterError("expected field OP value")
    field = tokens.pop(0)
    op = tokens.pop(0)
    if op not in _OPS:
        raise FilterError(f"unknown operator {op!r}")
    raw = tokens.pop(0)
    if raw == "null":
        value: Optional[str] = None
    else:
        value = raw
    return Comparison(field, op, value)


def _tokenize(expr: str) -> list[str]:
    tokens: list[str] = []
    i = 0
    n = len(expr)
    while i < n:
        ch = expr[i]
        if ch.isspace():
            i += 1
            continue
        if expr.startswith("&&", i):
            tokens.append("&&")
            i += 2
            continue
        matched_op = next((op for op in _OPS if expr.startswith(op, i)), None)
        if matched_op:
            tokens.append(matched_op)
            i += len(matched_op)
            continue
        if ch in ('"', "'"):
            quote = ch
            i += 1
            buf: list[str] = []
            while i < n and expr[i] != quote:
                if expr[i] == "\\" and i + 1 < n:
                    buf.append(expr[i + 1])
                    i += 2
                    continue
                buf.append(expr[i])
                i += 1
            if i >= n:
                raise FilterError("unterminated string")
            i += 1
            tokens.append("".join(buf))
            continue
        start = i
        while i < n and not expr[i].isspace() and not _starts_op(expr, i):
            i += 1
        word = expr[start:i]
        if word:
            tokens.append(word)
    return tokens


def _starts_op(expr: str, i: int) -> bool:
    if expr.startswith("&&", i):
        return True
    return any(expr.startswith(op, i) for op in _OPS)


class Comparison:
    def __init__(self, field: str, op: str, value: Optional[str]) -> None:
        self.field = field
        self.op = op
        self.value = value

    def eval(self, row: Row) -> bool:
        left = row.get(self.field) if row._selected is not None else row._raw_get(self.field)
        return _compare(left, self.op, self.value)


class And:
    def __init__(self, parts: list[Comparison]) -> None:
        self.parts = parts

    def eval(self, row: Row) -> bool:
        return all(part.eval(row) for part in self.parts)


FilterExpr = Union[Comparison, And]


def _compare(left: Any, op: str, right: Optional[str]) -> bool:
    if op == "==":
        if right is None:
            return left is None or left == ""
        return left is not None and str(left) == right
    if op == "!=":
        if right is None:
            return left is not None and left != ""
        return left is None or str(left) != right
    left_s = "" if left is None else str(left)
    if right is None:
        return False
    if op == "^=":
        return left is not None and left_s.startswith(right)
    if op == "!^=":
        return left is None or not left_s.startswith(right)
    return False

File: src/hql/expr.rs
//! HQL v0 filter expressions: == != ^= !^= with and / &&. null literal.

use crate::error::{Error, Result};
use crate::hql::row::{Row, Value};

const OPS: &[&str] = &["!^=", "==", "!=", "^="];

#[derive(Clone, Debug)]
pub enum FilterExpr {
    Comparison(Comparison),
    And(Vec<Comparison>),
}

#[derive(Clone, Debug)]
pub struct Comparison {
    pub field: String,
    pub op: String,
    pub value: Option<String>,
}

impl FilterExpr {
    pub fn eval(&self, row: &Row) -> bool {
        match self {
            FilterExpr::Comparison(c) => c.eval(row),
            FilterExpr::And(parts) => parts.iter().all(|part| part.eval(row)),
        }
    }
}

impl Comparison {
    pub fn eval(&self, row: &Row) -> bool {
        compare(&row.get(&self.field), &self.op, self.value.as_deref())
    }
}

pub fn parse_filter(expr: &str) -> Result<FilterExpr> {
    let mut tokens = tokenize(expr)?;
    if tokens.is_empty() {
        return Err(Error::Invalid("empty filter".into()));
    }
    parse_and(&mut tokens)
}

fn parse_and(tokens: &mut Vec<String>) -> Result<FilterExpr> {
    let mut parts = vec![parse_comparison(tokens)?];
    while !tokens.is_empty() {
        let join = tokens.remove(0);
        if join != "and" && join != "&&" {
            return Err(Error::Invalid(format!(
                "expected 'and' or '&&', got {join:?}"
            )));
        }
        parts.push(parse_comparison(tokens)?);
    }
    if parts.len() == 1 {
        Ok(FilterExpr::Comparison(parts.remove(0)))
    } else {
        Ok(FilterExpr::And(parts))
    }
}

fn parse_comparison(tokens: &mut Vec<String>) -> Result<Comparison> {
    if tokens.len() < 3 {
        return Err(Error::Invalid("expected field OP value".into()));
    }
    let field = tokens.remove(0);
    let op = tokens.remove(0);
    if !OPS.contains(&op.as_str()) {
        return Err(Error::Invalid(format!("unknown operator {op:?}")));
    }
    let raw = tokens.remove(0);
    let value = if raw == "null" { None } else { Some(raw) };
    Ok(Comparison { field, op, value })
}

fn tokenize(expr: &str) -> Result<Vec<String>> {
    let mut tokens = Vec::new();
    let chars: Vec<char> = expr.chars().collect();
    let n = chars.len();
    let mut i = 0;
    while i < n {
        let ch = chars[i];
        if ch.is_whitespace() {
            i += 1;
            continue;
        }
        if starts_with_at(expr, i, "&&") {
            tokens.push("&&".into());
            i += 2;
            continue;
        }
        if let Some(op) = OPS.iter().copied().find(|op| starts_with_at(expr, i, op)) {
            tokens.push(op.to_string());
            i += op.chars().count();
            continue;
        }
        if ch == '"' || ch == '\'' {
            let quote = ch;
            i += 1;
            let mut buf = String::new();
            while i < n && chars[i] != quote {
                if chars[i] == '\\' && i + 1 < n {
                    buf.push(chars[i + 1]);
                    i += 2;
                    continue;
                }
                buf.push(chars[i]);
                i += 1;
            }
            if i >= n {
                return Err(Error::Invalid("unterminated string".into()));
            }
            i += 1;
            tokens.push(buf);
            continue;
        }
        let start = i;
        while i < n && !chars[i].is_whitespace() && !starts_op_at(expr, i) {
            i += 1;
        }
        let word: String = chars[start..i].iter().collect();
        if !word.is_empty() {
            tokens.push(word);
        }
    }
    Ok(tokens)
}

fn starts_with_at(expr: &str, char_index: usize, needle: &str) -> bool {
    expr.chars()
        .skip(char_index)
        .collect::<String>()
        .starts_with(needle)
}

fn starts_op_at(expr: &str, char_index: usize) -> bool {
    if starts_with_at(expr, char_index, "&&") {
        return true;
    }
    OPS.iter().any(|op| starts_with_at(expr, char_index, op))
}

fn compare(left: &Value, op: &str, right: Option<&str>) -> bool {
    match op {
        "==" => match right {
            None => left.is_null() || matches!(left, Value::Str(s) if s.is_empty()),
            Some(right) => !left.is_null() && left.py_str() == right,
        },
        "!=" => match right {
            None => !left.is_null() && !matches!(left, Value::Str(s) if s.is_empty()),
            Some(right) => left.is_null() || left.py_str() != right,
        },
        "^=" => match right {
            None => false,
            Some(right) => !left.is_null() && left.py_str().starts_with(right),
        },
        "!^=" => match right {
            None => false,
            Some(right) => left.is_null() || !left.py_str().starts_with(right),
        },
        _ => false,
    }
}
